lib/fog/libvirt/models/compute/server.rb

```diff
@@ -257,11 +257,132 @@ def cloud_init_volume_name
         private
         attr_accessor :volumes_path
 
-        # This retrieves the ip address of the mac address
+        # This tests the library version before redefining the address
+        # method for this instance to use a method compatible with
+        # earlier libvirt libraries, or uses the dhcp method from more
+        # recent releases.
+        def addresses(service_arg=service, options={})
+          addresses_method = self.method(:addresses_dhcp)
+          version_regexp = Regexp.new("^Using library: libvirt \(.*\)$")
+          IO.popen("virsh version") do |p|
+            p.each_line do |l|
+              match = version_regexp.match(l)
+              if !match.nil?
+                # if we have a match, we'll have a subgroup
+                if Gem::Version.new(match[1]) <= Gem::Version.new('1.2.8')
+                  addresses_method = self.method(:addresses_ip_command)
+                end
+              end
+            end
+            status=Process.waitpid2(p.pid)[1].exitstatus
+            if status!=0
+              raise Fog::Errors::Error.new("The command 'virsh version' failed to execute with a clean exit code")
+            end
+          end
+          # replace current definition for this instance with correct one for
+          # detected libvirt to ensure detection is performed only once
+          (class << self; self; end).class_eval do
+            define_method(:addresses, addresses_method)
+          end
+          addresses(service_arg, options)
+        end
+
+        # This retrieves the ip address of the mac address using ip_command
         # It returns an array of public and private ip addresses
         # Currently only one ip address is returned, but in the future this could be multiple
         # if the server has multiple network interface
-        def addresses(service_arg=service, options={})
+        def addresses_ip_command(service_arg=service, options={})
```

Comment:
Note that this is a restoration of the original code that was replaced when switching to dhcp_leases.